### PR TITLE
Add nfs and finding kernel

### PR DESCRIPTION
**Pull request by PatrykHub**

### CODE REVIEW

File: mcda/electre/outranking.py [hunk 1]
```diff
@@ -486,3 +486,117 @@ def distillation(
...
+                alt_name_a
+                for alt_name_a in crisp_outranking_table.index
+                if crisp_outranking_table.loc[alt_name_a][alt_name_b] != 0
+            ]
+            for alt_name_b in crisp_outranking_table.index
+        }
+    )
+
+
+def strongly_connected_components(graph: pd.Series) -> List:
```
Comment (agak33) [inline mcda/electre/outranking.py:504]: Please add more information about the returned list (what is inside?)

File: mcda/electre/outranking.py [hunk 2]
```diff
@@ -486,3 +486,117 @@ def distillation(
...
+    for vertices in strongly_connected_components(graph):
+        if len(vertices) == 1:
+            continue
+        aggregated = ", ".join(str(v) for v in vertices)
+        new_connections = list(
+            set([v for key in vertices for v in graph[key] if v not in vertices])
+        )
+        graph.drop(labels=vertices)
+        for key in new_graph:
+            for vertex in new_graph[key][:]:
```
Comment (agak33) [inline mcda/electre/outranking.py:550]: `[:]` is unecessary here

File: mcda/electre/outranking.py [hunk 3]
```diff
@@ -486,3 +486,117 @@ def distillation(
...
+        }
+    )
+
+
+def strongly_connected_components(graph: pd.Series) -> List:
+    index_counter = [0]
+    stack, result = [], []
+    lowlink, index = {}, {}
+
+    def _strong_connect(node):
```
Comment (agak33) [inline mcda/electre/outranking.py:509]: Add a comment about what this function does

File: mcda/electre/outranking.py [hunk 4]
```diff
@@ -486,3 +486,117 @@ def distillation(
...
+
+        if lowlink[node] == index[node]:
+            connected_component = []
+
+            while True:
+                successor = stack.pop()
+                connected_component.append(successor)
+                if successor == node:
+                    break
+            result.append(connected_component[:])
```
Comment (agak33) [inline mcda/electre/outranking.py:531]: `[:]` is unecessary here

File: mcda/electre/outranking.py [hunk 5]
```diff
@@ -486,3 +486,117 @@ def distillation(
...
+
+def aggregate(graph: pd.Series) -> pd.Series:
+    new_graph = graph.copy()
+    for vertices in strongly_connected_components(graph):
+        if len(vertices) == 1:
+            continue
+        aggregated = ", ".join(str(v) for v in vertices)
+        new_connections = list(
+            set([v for key in vertices for v in graph[key] if v not in vertices])
+        )
```
Comment (agak33) [inline mcda/electre/outranking.py:547]: Couldn't it be a little bit simplified? XD

File: mcda/electre/outranking.py [hunk 6]
```diff
@@ -486,3 +486,117 @@ def distillation(
...
+            for vertex in new_graph[key][:]:
+                if vertex in vertices:
+                    new_graph[key].remove(vertex)
+                    if aggregated not in new_graph[key]:
+                        new_graph[key].append(aggregated)
+        new_graph[aggregated] = new_connections
+    return new_graph
+
+
+def find_vertices_without_predecessor(graph: pd.Series) -> List:
```
Comment (agak33) [inline mcda/electre/outranking.py:559]: same as above - add more info about returned list.

File: mcda/electre/outranking.py [hunk 7]
```diff
@@ -486,3 +486,117 @@ def distillation(
...
+    vertices_with_preedecessor = list(set([v for key in graph for v in graph[key]]))
+    return [vertex for vertex in graph if vertex not in vertices_with_preedecessor]
+
+
+def find_kernel(crisp_outranking_table: pd.DataFrame) -> List:
```
Comment (agak33) [inline mcda/electre/outranking.py:564]: same as above

**Review by a-pralat — APPROVED** [review]
Looks fine after these changes. Please resolve conflicts and merge this @PatrykHub

**Pull request merged**